### DISCUSSION TITLE
Calculate the 3D coordinates of objects detected in the image

### DISCUSSION THREAD

**Discussion by BlueWolf7** [ultralytics/ultralytics#10371] [Q&A]
I'm planning to use a single camera  to detect various points of interest such as banks and pharmacies (specifically, to detect their logos). I have already trained a YOLOv8 model for this purpose and aim to calculate their 3D coordinates . Can anyone  help me with a code that can calculate the 3D coordinates of detected objects in the image using a single camera? I'm considering treating two consecutive images from the camera as if they were captured by two separate cameras, thereby making stereo vision applicable. Additionally, the coordinates of the camera are always known via GPS

**Comment by pderrenger** [comment]
Calculating 3D coordinates from a single camera setup can be quite challenging, as depth information is inherently missing from a 2D image. However, your approach of using two consecutive images as a stereo vision system is an interesting workaround! 🧐

For this purpose, you would first need to detect objects in both images using your YOLOv8 model. Once detected, you can use feature matching techniques between the two images to estimate the relative position of these objects. OpenCV is a fantastic library that can help with feature matching and stereo vision calculations.

Here is a basic outline of what the code could look like using OpenCV in Python, assuming you've got both images (`image1`, `image2`) and object detections from YOLOv8:

```python
import cv2
import numpy as np

# Load your images
image1 = cv2.imread('path_to_image1.jpg')
image2 = cv2.imread('path_to_image2.jpg')

# Assume `detections1` and `detections2` hold YOLO results for each image
# You need to extract keypoints for each object detection in both images
# For simplicity, let's assume (x, y) is the center of a detected object

# Find keypoints and descriptors with SIFT
sift = cv2.SIFT_create()
keypoints1, descriptors1 = sift.detectAndCompute(image1, None)
keypoints2, descriptors2 = sift.detectAndCompute(image2, None)

# FLANN based Matcher
FLANN_INDEX_KDTREE = 1
index_params = dict(algorithm = FLANN_INDEX_KDTREE, trees = 5)
search_params = dict(checks=50)   # or pass an empty dictionary

flann = cv2.FlannBasedMatcher(index_params,search_params)
matches = flann.knnMatch(descriptors1,descriptors2,k=2)

# Filter matches using the Lowe's ratio test
good_matches = []
for m,n in matches:
    if m.distance < 0.7*n.distance:
        good_matches.append(m)

# Assuming you have the camera intrinsic parameters and the relative pose
# between image1 and image2, you can now triangulate points to get 3D coordinates
# (This part highly depends on your camera setup and calibration)

# Note: This is a simplified example, you would need to integrate your detections
# and possibly adjust for the camera's intrinsic and extrinsic parameters.
```

This is a very high-level and simplified example. Real-world application will require careful camera calibration to determine intrinsic (e.g., focal length, principal point) and extrinsic (position, orientation) parameters, handling of matching points to detected objects, and substantial error checking. For true 3D coordinate computation, consider a more comprehensive stereo vision setup or additional sensors that can provide depth information directly.
